File: backend/services/order_service.py
from models.user import User
from models.order import Order
from models.order import OrderItem
from models.dish import Dish
from services.user_service import apply_warning, update_vip_status
import datetime

def create_order(customer_id, items):
    customer = User.objects(id=customer_id).first()
    if not customer or customer.status != "Active":
        return {"error": "Invalid or inactive customer."}, 400
    
    order_items = []
    for item in items:
        dish = Dish.objects(id=item['dish_id']).first()
        if not dish or not dish.isAvailable:
            return {"error": f"Dish with id {item['dish_id']} not found or unavailable."}, 400
        order_item = OrderItem(
            dish=dish,
            quantity=item["quantity"],
            price=dish.price
        )
        order_items.append(order_item)
    
    order = Order(
        customer=customer,
        items=order_items,
        status="PendingPayment",
    )
    order.save()
    order.calculate_total_price()
    
    return {
        "order_id": str(order.id),
        "original_price": order.original_price,
        "discount_applied": order.discount_applied,
        "final_price": order.final_price
    }, 201

def confirm_order(order_id):
    order = Order.objects(id=order_id).first()
    if not order:
        return {"error": "Order not found."}, 404
    
    customer = order.customer
    
    if customer.balance < order.final_price:
        order.set_status("Rejected_Insufficient_Funds")
        apply_warning(customer.id, "Insufficient funds for order payment.")
        return {
            "error": "Insufficient funds.",
            "balance": customer.balance,
        }, 400
    
    customer.balance -= order.final_price
    customer.totalSpent += order.final_price
    customer.orderCount += 1
    customer.save()

    order.increment_dish_order_counts()
    update_vip_status(customer.id)
    order.set_status("Paid")
    order.set_status("Queued_For_Preparation")
    return {
        "message": "Order confirmed and paid successfully. Order sent to kitchen.",
        "order_id": str(order.id)
    }, 200

def get_customer_orders(customer_id):
    orders = Order.objects(customer=customer_id).order_by('-created_at')
    return [
        {"id": str(order.id),
         "status": order.status,
         "final_price": order.final_price,
         "created_at": order.created_at
         } for order in orders
    ]

File: backend/services/order_service.js
const User = require('../models/User');
const Order = require('../models/Order');
const Dish = require('../models/Dish');
const { applyWarning, updateVipStatus } = require('./user_service');

// create a new order
async function createOrder(customerId, items) {
  const customer = await User.findById(customerId);
  
  if (!customer || !["Customer", "VIP"].includes(customer.role)) {
    return { error: "Invalid customer. Must be logged in as Customer or VIP." }, 400;
  }
  
  if (customer.status !== "Active") {
    return { error: "Account is not active." }, 400;
  }
  
  if (customer.balance <= 0) {
    return { error: "Insufficient balance. Please deposit funds first." }, 400;
  }
  
  const orderItems = [];
  for (const item of items) {
    const dish = await Dish.findById(item.dish_id);
    if (!dish || !dish.is_available) {
      return { error: `Dish with id ${item.dish_id} not found or unavailable.` }, 400;
    }
    orderItems.push({
      dish: dish._id,
      quantity: item.quantity,
      price: dish.price
    });
  }
  
  const order = new Order({
    customer: customer._id,
    items: orderItems,
    status: "PendingPayment"
  });
  
  await order.save();
  order.customer = customer; // attach customer object for vip check
  await order.calculate_total_price();
  
  return {
    order_id: order._id.toString(),
    original_price: order.original_price,
    discount_applied: order.discount_applied,
    final_price: order.final_price,
    customer_balance: customer.balance
  }, 201;
}

// confirm and pay for an order
async function confirmOrder(orderId) {
  const order = await Order.findById(orderId).populate('customer');
  if (!order) {
    return { error: "Order not found." }, 404;
  }
  
  const customer = order.customer;
  
  if (customer.balance < order.final_price) {
    await order.set_status("Rejected_Insufficient_Funds");
    await applyWarning(customer._id.toString(), "Order attempt with insufficient funds");
    
    return {
      error: "Insufficient funds. Order rejected. Warning applied.",
      balance: customer.balance,
      required: order.final_price
    }, 400;
  }
  
  customer.balance -= order.final_price;
  customer.totalSpent += order.final_price;
  customer.orderCount += 1;
  await customer.save();
  
  await order.increment_dish_order_counts();
  
  const vipUpgraded = await updateVipStatus(customer._id.toString());
  
  await order.set_status("Paid");
  await order.set_status("Queued_For_Preparation");
  
  let message = "Order confirmed and paid successfully. Order sent to kitchen.";
  if (vipUpgraded) {
    message += " Congratulations! You've been upgraded to VIP status!";
  }
  
  return {
    message,
    order_id: order._id.toString(),
    new_balance: customer.balance,
    vip_upgraded: vipUpgraded
  }, 200;
}

// get orders for a customer
async function getCustomerOrders(customerId) {
  const orders = await Order.find({ customer: customerId })
    .populate('items.dish')
    .sort({ created_at: -1 });
  
  return orders.map(order => ({
    id: order._id.toString(),
    status: order.status,
    final_price: order.final_price,
    original_price: order.original_price,
    discount_applied: order.discount_applied,
    created_at: order.created_at,
    items: order.items.map(item => ({
      dish_name: item.dish.name,
      quantity: item.quantity,
      price: item.price
    }))
  }));
}

module.exports = {
  createOrder,
  confirmOrder,
  getCustomerOrders
};
